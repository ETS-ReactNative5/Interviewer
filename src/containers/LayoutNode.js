/* eslint-disable */

import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { first } from 'lodash';
import { Node } from 'network-canvas-ui';
import { draggable, selectable } from '../behaviours';
import { actionCreators as networkActions } from '../ducks/modules/network';

const label = node => node.nickname;

const EnhancedNode = draggable(selectable(Node));

class LayoutNode extends PureComponent {
  onDropped = (...args) => {
    this.onMove(...args);
    this.props.onDropped();
  }

  onMove = (hits, { x, y }) => {
    const { layout, updateNode, node: { uid } } = this.props;
    const hitbox = first(hits);

    // Calculate x/y position as a decimal within the hitbox
    const relativeCoords = {
      type: 'layout',
      x: (x - hitbox.x) / hitbox.width,
      y: (y - hitbox.y) / hitbox.height,
    };

    updateNode({ uid, [layout]: relativeCoords });
  }

  onSelected = () => {
    this.props.onSelected(this.props.node);
  }

  render() {
    const {
      node,
      draggableType,
      selected,
      canDrag,
      canSelect,
      layout,
      areaWidth,
      areaHeight,
    } = this.props;

    const { x, y } = node[layout];

    const styles = {
      left: 0,
      top: 0,
      transform: `translate(calc(${x * areaWidth}px - 50%), calc(${y * areaHeight}px - 50%))`,
    };

    return (
      <div
        className="node-layout__node"
        style={styles}
      >
        <EnhancedNode
          label={label(node)}
          draggableType={draggableType}
          onDropped={this.onDropped}
          onMove={this.onMove}
          onSelected={this.onSelected}
          selected={selected}
          canDrag={canDrag}
          canSelect={canSelect}
          animate={false}
          {...node}
        />
      </div>
    );
  }
}

LayoutNode.propTypes = {
  onDropped: PropTypes.func.isRequired,
  updateNode: PropTypes.func.isRequired,
  onSelected: PropTypes.func.isRequired,
  layout: PropTypes.string.isRequired,
  node: PropTypes.object.isRequired,
  draggableType: PropTypes.string.isRequired,
  canDrag: PropTypes.bool,
  canSelect: PropTypes.bool,
  selected: PropTypes.bool,
};

LayoutNode.defaultProps = {
  canDrag: false,
  canSelect: false,
  selected: false,
};

function mapDispatchToProps(dispatch) {
  return {
    updateNode: bindActionCreators(networkActions.updateNode, dispatch),
  };
}

export { LayoutNode as LayoutNodePure };
export default connect(null, mapDispatchToProps)(LayoutNode);
