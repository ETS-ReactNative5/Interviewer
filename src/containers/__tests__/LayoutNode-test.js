/* eslint-env jest */

import React from 'react';
import { shallow } from 'enzyme';
import { LayoutNodePure as LayoutNode } from '../LayoutNode';

const layout = 'foo';

const mockProps = {
  onDropped: () => {},
  updateNode: () => {},
  onSelected: () => {},
  layout,
  node: { [layout]: { x: 0.77, y: 0.2 } },
  areaWidth: 100,
  areaHeight: 40,
  draggableType: 'bar',
};

describe('<LayoutNode />', () => {
  it('renders ok', () => {
    const component = shallow(<LayoutNode {...mockProps} />);

    expect(component).toMatchSnapshot();
  });

  it('positions using translate', () => {
    const component = shallow(<LayoutNode {...mockProps} />);

    expect(component.prop('style')).toEqual(
      {
        left: 0,
        top: 0,
        transform: 'translate(calc(77px - 50%), calc(8px - 50%))',
      },
    );
  });
});
