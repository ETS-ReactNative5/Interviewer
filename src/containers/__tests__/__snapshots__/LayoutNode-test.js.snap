// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LayoutNode /> renders ok 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="node-layout__node"
    style={
        Object {
            "left": 0,
            "top": 0,
            "transform": "translate(calc(77px - 50%), calc(8px - 50%))",
          }
    }
>
    <Connect(Draggable)
        animate={false}
        canDrag={false}
        canSelect={false}
        draggableType="bar"
        foo={
            Object {
                "x": 0.77,
                "y": 0.2,
              }
        }
        label={undefined}
        onDropped={[Function]}
        onMove={[Function]}
        onSelected={[Function]}
        selected={false}
    />
</div>,
  "nodes": Array [
    <div
      className="node-layout__node"
      style={
            Object {
                  "left": 0,
                  "top": 0,
                  "transform": "translate(calc(77px - 50%), calc(8px - 50%))",
                }
      }
>
      <Connect(Draggable)
            animate={false}
            canDrag={false}
            canSelect={false}
            draggableType="bar"
            foo={
                  Object {
                        "x": 0.77,
                        "y": 0.2,
                      }
            }
            label={undefined}
            onDropped={[Function]}
            onMove={[Function]}
            onSelected={[Function]}
            selected={false}
      />
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 1,
      "_context": Object {},
      "_currentElement": <LayoutNode
        areaHeight={40}
        areaWidth={100}
        canDrag={false}
        canSelect={false}
        draggableType="bar"
        layout="foo"
        node={
                Object {
                        "foo": Object {
                          "x": 0.77,
                          "y": 0.2,
                        },
                      }
        }
        onDropped={[Function]}
        onSelected={[Function]}
        selected={false}
        updateNode={[Function]}
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": LayoutNode {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "onDropped": [Function],
        "onMove": [Function],
        "onSelected": [Function],
        "props": Object {
          "areaHeight": 40,
          "areaWidth": 100,
          "canDrag": false,
          "canSelect": false,
          "draggableType": "bar",
          "layout": "foo",
          "node": Object {
            "foo": Object {
              "x": 0.77,
              "y": 0.2,
            },
          },
          "onDropped": [Function],
          "onSelected": [Function],
          "selected": false,
          "updateNode": [Function],
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="node-layout__node"
          style={
                    Object {
                              "left": 0,
                              "top": 0,
                              "transform": "translate(calc(77px - 50%), calc(8px - 50%))",
                            }
          }
>
          <Connect(Draggable)
                    animate={false}
                    canDrag={false}
                    canSelect={false}
                    draggableType="bar"
                    foo={
                              Object {
                                        "x": 0.77,
                                        "y": 0.2,
                                      }
                    }
                    label={undefined}
                    onDropped={[Function]}
                    onMove={[Function]}
                    onSelected={[Function]}
                    selected={false}
          />
</div>,
        "_debugID": 2,
        "_renderedOutput": <div
          className="node-layout__node"
          style={
                    Object {
                              "left": 0,
                              "top": 0,
                              "transform": "translate(calc(77px - 50%), calc(8px - 50%))",
                            }
          }
>
          <Connect(Draggable)
                    animate={false}
                    canDrag={false}
                    canSelect={false}
                    draggableType="bar"
                    foo={
                              Object {
                                        "x": 0.77,
                                        "y": 0.2,
                                      }
                    }
                    label={undefined}
                    onDropped={[Function]}
                    onMove={[Function]}
                    onSelected={[Function]}
                    selected={false}
          />
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <LayoutNode
    areaHeight={40}
    areaWidth={100}
    canDrag={false}
    canSelect={false}
    draggableType="bar"
    layout="foo"
    node={
        Object {
            "foo": Object {
              "x": 0.77,
              "y": 0.2,
            },
          }
    }
    onDropped={[Function]}
    onSelected={[Function]}
    selected={false}
    updateNode={[Function]}
/>,
}
`;
